Library/Homebrew/utils/curl.rb

```diff
@@ -19,6 +20,8 @@ def curl(*args)
 end
 
 def curl_output(*args)
-  curl_args = curl_args(args) - ["--fail"]
-  Utils.popen_read_text(*curl_args)
+  curl_args = curl_args(args) - ["--fail", "--silent"]
+  Open3.popen3(*curl_args) do |_, stdout, stderr, status|
```

Comment:
I'd probably lean towards using the new code for Ruby 2.0+ and falling back to `Utils.popen_read_text` for older Rubies. That way most of our users will benefit from an improved error handling and we can rip out the legacy code path as soon as we enforce Ruby 2.0+. A very rough and untested sketch (that would also require some minor adjustments on the call site):

```ruby
def curl_output(*args)
  curl_args = curl_args(args) - ["--fail"]
  if RUBY_VERSION.split(".").first.to_i >= 2
    result = Open3.popen3(*curl_args) do |_, stdout, stderr, status|
      [stdout.read, stderr.read, status]
    end
    result[0..1] + [result.last.success?]
  else
    output = Utils.popen_read_text(*curl_args)
    [output, "(Curl error message unavailable.)", $?.success?]
  end
end
```